file.go

```diff
@@ -0,0 +1,136 @@
+	}
+	if resp.Header.Get("Content-Type") != "application/zip" {
+		return "", newError(500, fmt.Sprintf("Failed to download file at the url %s. Expected HTTP Response's \"Content-Type\" header to be \"application/zip\", but was \"%s\"", url, resp.Header.Get("Content-Type")))
+	}
+
+	// Copy the contents of the downloaded file to our empty file
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return "", newErr(err)
+	}
+
+	return filepath.Join(tempDir, "repo.zip"), newEmptyError()
+}
+
+// extractFiles decompresses the file at zipFileAbsPath and moves only those files under filesToExtractFromZipPath to localPath
+func extractFiles(zipFilePath, filesToExtractFromZipPath, localPath string) error {
+
+	// Open the zip file for reading.
+	r, err := zip.OpenReader(zipFilePath)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// pathPrefix represents the portion of the local file path we will ignore when copying the file to localPath
+	// E.g. full path = fetch-test-public-0.0.3/folder/file1.txt
+	//      path prefix = fetch-test-public-0.0.3
+	//      file that will eventually get written = <localPath>/folder/file1.txt
+
+	// By convention, the first file in the zip file is the top-level directory
+	pathPrefix := r.File[0].Name
+
+	// Add the path from which we will extract files to the path prefix so we can exclude the appropriate files
+	pathPrefix = filepath.Join(pathPrefix, filesToExtractFromZipPath)
+
+	// Iterate through the files in the archive,
+	// printing some of their contents.
+	for _, f := range r.File {
+
+		// If the given file is in the filesToExtractFromZipPath, proceed
+		if strings.Index(f.Name, pathPrefix) == 0 {
+
+			// Read the contents of the file in the .zip file
+			readCloser, err := f.Open()
+			if err != nil {
+				return fmt.Errorf("Failed to open file %s: %s", f.Name, err)
+			}
+			defer readCloser.Close()
+
+
+			if f.FileInfo().IsDir() {
+				// Create a directory
+				os.MkdirAll(filepath.Join(localPath, strings.TrimPrefix(f.Name, pathPrefix)), 0777)
+			} else {
+				// Create a new empty file
+				fmt.Printf("Writing file %s\n", filepath.Join(localPath, strings.TrimPrefix(f.Name, pathPrefix)))
+				file, err := os.Create(filepath.Join(localPath, strings.TrimPrefix(f.Name, pathPrefix)))
+				if err != nil {
+					return fmt.Errorf("Failed to create new file: %s", err)
+				}
+				defer file.Close()
+
+				// Copy the contents to it
+				_, err = io.Copy(file, readCloser)
+				if err != nil {
+					return fmt.Errorf("Failed to copy file: %s", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// getZipFileName extracts the zip file name from the absolute file path of the zip file
+// It returns the name without the .zip suffix
+func getZipFileName(zipFilePath string) string {
+	exploded := strings.Split(zipFilePath, "/")
+	return strings.TrimSuffix(exploded[len(exploded)-1], ".zip")
```

Comment:
File paths can be very tricky due to `..`, `.`, and other weirdness. The Go `path` package may be a safer bet: e.g. `return strings.TrimSuffix(path.Base(zipFilePath), path.Ext(zipFilePath))`.